### PR TITLE
Numpy string coding

### PR DESCRIPTION
**Pull request by znicholls**
Fixes handling of numpy string types in coding

- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by znicholls** [comment]
So far I've just added a single test which fails. I don't think the test should fail although I'm not sure what the `np.str_` type actually is so maybe this isn't a bug? Help/advice greatly appreciated.

**Comment by shoyer** [comment]
> So far I've just added a single test which fails. I don't think the test should fail although I'm not sure what the `np.str_` type actually is so maybe this isn't a bug? Help/advice greatly appreciated.

What problem are you trying to solve here?

This vlen string stuff is an internal API that isn't really intended for use outside Xarray.

**Comment by znicholls** [comment]
> What problem are you trying to solve here?

Somehow I ended up with `np.str_` in a pandas dataframe, which then exploded when I converted to xarray and attempted to save as netCDF. Minimal example below.

```python
import numpy as np
import pandas as pd
import scmdata


# I don't know how the strings ended up being np.str_....
scenarios = [np.str_(v) for v in ["scenario_a", "scenario_b", "scenario_c"]]
years = range(2015, 2100 + 1)
tdf = pd.DataFrame(
    data=np.random.random((len(scenarios), len(years))),
    columns=years,
    index=scenarios,
)
tdf.index.name = "scenario"
tdf.columns.name = "year"
tdf = tdf.stack()
tdf.name = "tas"

txr = tdf.to_xarray()
# raises error
txr.to_netcdf("test.nc")
```

**Comment by znicholls** [comment]
@shoyer any further thoughts on this now that the scope is clearer?

**Comment by shoyer** [comment]
I agree, this should totally work. It's not obvious to me how to best fix it, though.

**Comment by znicholls** [comment]
> I agree, this should totally work. It's not obvious to me how to best fix it, though.

I assume it's not as trivial as just changing e.g. https://github.com/pydata/xarray/blob/f9a535c130615fe81379d785dffcc6c814fc1126/xarray/coding/strings.py#L32 to also know about `np.str_`?

**Comment by shoyer** [comment]
I think the issue must be somewhere around this line, where xarray attempts to infer a dtype for object arrays:
https://github.com/pydata/xarray/blob/f9a535c130615fe81379d785dffcc6c814fc1126/xarray/conventions.py#L215

**Comment by znicholls** [comment]
I tried pushing a fix. It's unclear to me whether the change should be in how the dtypes are inferred (given that the inference code seems to do what it is meant to...) or whether `is_unicode_dtype` simply needs to be updated to know about `np.str_` (which is the fix I just tried).

**Comment by shoyer** [comment]
My suggestion is that either `_infer_dtype` (https://github.com/pydata/xarray/blob/f9a535c130615fe81379d785dffcc6c814fc1126/xarray/conventions.py#L160-L161
) or the underlying `create_vlen_dtype` should be updated, so it never puts `np.str_` inside a custom vlen dtype. Instead, we should normalize `element_type` to always be `str` or `bytes` inside the vlen dtype.

**Comment by shoyer** [comment]
To add a bit more clarification: the vlen dtype should correspond to an HDF5/netCDF4 compatible data-type, like a variable length string or bytes. `np.str_` is just a NumPy variant of `str`, so the correct dtype is `create_vlen_dtype(str)`.

**Comment by znicholls** [comment]
Something like 59ed7d5?

**Comment by znicholls** [comment]
`xarray/tests/test_backends.py::test_open_fsspec` appears to fail because of the release of ffspec 2021.7.0 so this PR will probably have to wait until a fix for that is added (presumably elsewhere to keep the changes clear).

**Comment by znicholls** [comment]
Ignoring failing CI due to fsspec (see https://github.com/pydata/xarray/pull/5615#issuecomment-881679900)

**Comment by znicholls** [comment]
@shoyer can I bother you again now that CI is passing please?

**Comment by znicholls** [comment]
@lewisjarednz fyi

File: xarray/tests/test_coding_strings.py [hunk 1]
```diff
@@ -29,6 +36,33 @@ def test_vlen_dtype() -> None:
...
+@pytest.mark.parametrize("numpy_str_type", (np.str_, np.bytes_))
+def test_numpy_subclass_handling(numpy_str_type) -> None:
+    with pytest.raises(TypeError, match="unsupported type for vlen_dtype"):
+        strings.create_vlen_dtype(numpy_str_type)
+
+
+@requires_netCDF4
+@pytest.mark.parametrize("str_type", (str, np.str_))
+def test_write_file_from_np_str(str_type) -> None:
+    # should be moved elsewhere probably
```
Comment (shoyer) [inline xarray/tests/test_coding_strings.py:48]: Please move this to test_backends.py, and include a comment linking back to the GitHub issue (for context).

**Review by shoyer — COMMENTED** [review]
Looks great, thanks! Please move the test, then we can merge this

File: xarray/tests/test_coding_strings.py (hunk 1 above)
Comment (znicholls) [reply xarray/tests/test_coding_strings.py:48]: 👍

**Comment by Illviljan** [comment]
Looks good to me, nice work!

**Comment by Illviljan** [comment]
Thanks @znicholls!

**Pull request merged**